### PR TITLE
Minor Liquidatable refactor

### PR DESCRIPTION
**Pull request by mrice32**
This PR slightly refactors Liquidatable so that we track both the last used index and the liquidations in an array rather than using separate data structures.

All the methods continue to work similarly as they did before, except that when the array is accessed out of bounds, solidity automatically throws rather than giving an uninitialized struct.

This also includes a slight rewriting of some of the modifiers, so we don't repeatedly call the same method.

### CODE REVIEW

File: core/contracts/financial_templates/Liquidatable.sol [hunk 1]
```diff
@@ -79,9 +79,7 @@ contract Liquidatable is PricelessPositionManager {
     */
 
     // Liquidations are unique by ID per sponsor
-    mapping(address => mapping(uint => LiquidationData)) public liquidations;
-    // Keeps track of last used liquidation ID per sponsor
-    mapping(address => uint) public sponsorLiquidationIndex;
+    mapping(address => LiquidationData[]) public liquidations;
```
Comment (ptare) [inline core/contracts/financial_templates/Liquidatable.sol]: What effect does this change have on the running time of `delete liquidations[sponsor][id]` on line 352?

Comment (mrice32) [reply core/contracts/financial_templates/Liquidatable.sol]: Good point -- thanks for pointing this out.

I just looked into the [solidity docs on this](https://solidity.readthedocs.io/en/v0.5.13/types.html#delete), and by my reading, this should do what we expect/want: zero out the struct without touching the length of the array.

File: core/contracts/financial_templates/Liquidatable.sol [hunk 2]
```diff
@@ -107,21 +105,22 @@ contract Liquidatable is PricelessPositionManager {
...
     // i.e. once a dispute has been requested, the liquidation's expiry ceases to matter
     modifier onlyPostExpiryOrPostDispute(uint id, address sponsor) {
+        LiquidationData storage liquidation = _getLiquidationData(sponsor, id);
+        Status state = liquidation.state;
         require(
-            (_getLiquidationData(sponsor, id).state == Status.DisputeSucceeded) ||
-                (_getLiquidationData(sponsor, id).state == Status.DisputeFailed) ||
-                ((_getLiquidationData(sponsor, id).expiry <= getCurrentTime()) &&
-                    (_getLiquidationData(sponsor, id).state == Status.PreDispute)),
+            (state == Status.DisputeSucceeded) ||
```
Comment (nicholaspai) [inline core/contracts/financial_templates/Liquidatable.sol]: Much cleaner than before !

File: core/contracts/financial_templates/Liquidatable.sol [hunk 3]
```diff
@@ -170,40 +169,40 @@ contract Liquidatable is PricelessPositionManager {
...
-        // This should reduce totalTokensOutstanding and lockedCollateral, and also withdrawal request amount?
-        positionToLiquidate.tokensOutstanding = FixedPoint.fromUnscaledUint(0);
-        positionToLiquidate.collateral = FixedPoint.fromUnscaledUint(0);
-
-        // Set parameters for new liquidation
-        newLiquidation.expiry = getCurrentTime() + liquidationLiveness;
-        newLiquidation.liquidator = msg.sender;
-        newLiquidation.state = Status.PreDispute;
+        // UUID is the index of the LiquidationData that was just pushed into the array, which is length - 1.
+        uuid = newLength.sub(1);
```
Comment (nicholaspai) [inline core/contracts/financial_templates/Liquidatable.sol]: Perhaps you comment on this above from @ptare 's comment, but have you empirically verified that `delete` does not shift around the array such that the `uuid`'s would have to change as well?

File: core/contracts/financial_templates/Liquidatable.sol (hunk 1 above)
Comment (nicholaspai) [reply core/contracts/financial_templates/Liquidatable.sol]: I think you are right based on the line 

> It is important to note that delete a really behaves like an assignment to a, i.e. it stores a new object in a

[Another reference to this situation](https://ethereum.stackexchange.com/questions/1527/how-to-delete-an-element-at-a-certain-index-in-an-array)

File: core/contracts/financial_templates/Liquidatable.sol [hunk 4]
```diff
@@ -387,6 +386,5 @@ contract Liquidatable is PricelessPositionManager {
         returns (LiquidationData storage liquidation)
     {
         liquidation = liquidations[sponsor][uuid];
-        require(liquidation.liquidator != address(0), "Liquidation does not exist: liquidator address is not set");
```
Comment (nicholaspai) [inline core/contracts/financial_templates/Liquidatable.sol]: `PricelessPositionManager` has a check in `_getPositionData` that the position is valid, which is why I included this line. Do you think its ok to just allow caller of `_getLiquidationData` to revert if it tries to perform operations on an invalid Liquidation?

The tradeoff might be more obscure revert messages.

File: core/contracts/financial_templates/Liquidatable.sol [hunk 5]
```diff
@@ -387,6 +386,5 @@ contract Liquidatable is PricelessPositionManager {
         returns (LiquidationData storage liquidation)
     {
         liquidation = liquidations[sponsor][uuid];
```
Comment (nicholaspai) [inline core/contracts/financial_templates/Liquidatable.sol]: Looking at the test now (beginning line 214), it seems that this line will throw. Any point in wrapping it in a require statement so we can make error messages more clear?

File: core/contracts/financial_templates/Liquidatable.sol (hunk 3 above)
Comment (mrice32) [reply core/contracts/financial_templates/Liquidatable.sol]: I just verified this behavior in Remix -- delete on an element does not modify the length or any other element in the array (including other elements' positions).

File: core/contracts/financial_templates/Liquidatable.sol (hunk 4 above)
Comment (mrice32) [reply core/contracts/financial_templates/Liquidatable.sol]: Ah, I made a bad assumption here -- that the only invalid liquidations were ones that were uninitialized (and therefore, past the end of the array). Once a liquidation is fully resolved, we delete it, so we'd want to protect users from accessing those as well. 

I'll add this revert back. Thanks for pointing this out. Definitely an oversight.

File: core/contracts/financial_templates/Liquidatable.sol (hunk 5 above)
Comment (mrice32) [reply core/contracts/financial_templates/Liquidatable.sol]: Will add a require here.

File: core/contracts/financial_templates/Liquidatable.sol (hunk 4 above)
Comment (mrice32) [reply core/contracts/financial_templates/Liquidatable.sol]: Done.

File: core/contracts/financial_templates/Liquidatable.sol (hunk 5 above)
Comment (mrice32) [reply core/contracts/financial_templates/Liquidatable.sol]: Done.

**Comment by mrice32** [comment]
@ptare @nicholaspai do y’all have any follow up comments or are y’all happy with this?

**Comment by nicholaspai** [comment]
I'm good with this after reviewing your latest change to `_getLiquidationData`

> @ptare @nicholaspai do y’all have any follow up comments or are y’all happy with this?

**Pull request merged**